src/org/zaproxy/zap/extension/ascanrulesAlpha/HttpOnlySite.java

```diff
@@ -156,58 +152,70 @@ public void scan() {
 		}
 
 		try {
-			String record = "http"; //If any redirection is https record will be changed to https
+			String scheme = "http"; //If any redirection is https scheme will be changed to https
 			int count = 0;
 			while(count < 10) {
+				if (isStop()) {
+					if (log.isDebugEnabled()) {
+						log.debug("Scanner "+getName()+" Stopping.");
+					}
+					return;
+				}
 				sendAndReceive(newRequest, false);
-				System.out.println("\n\nThe request was succesful and it was\n"+newRequest.getRequestHeader().toString());
-				System.out.println("\n\nA response was recieved and it was\n"+newRequest.getResponseHeader().toString());
-				System.out.println("\n\nThe status code is: "+newRequest.getResponseHeader().getStatusCode());
-				String loc = newRequest.getResponseHeader().LOCATION;
-				System.out.println("\n\nThe location is: "+newRequest.getResponseHeader().getHeader(loc));
 				int status = newRequest.getResponseHeader().getStatusCode();
-				String redirect = newRequest.getResponseHeader().getHeader(loc);
-				if(status == HttpStatusCode.MOVED_PERMANENTLY || status == HttpStatusCode.FOUND) {
-					String path = newRequest.getRequestHeader().getURI().getPath();
-					if(redirect.contains("https")) {
-						String newHost = redirect.substring(8, redirect.length());
-					    newRequest.getRequestHeader().setURI(new URI("https", null, newHost, 443, path));
-					    record = "https";
-					    count++;
-					} else if (redirect.contains("http")) {
-						raiseAlert(newRequest);
-						break;
+				if(HttpStatusCode.isRedirection(status)) {
+					String loc = HttpResponseHeader.LOCATION;
+					String redirect = newRequest.getResponseHeader().getHeader(loc);
+					if(!redirect.isEmpty() && redirect != null) {
+						URI oldURI = newRequest.getRequestHeader().getURI();
+						try {
+							newRequest.getRequestHeader().setURI(new URI(oldURI, redirect, true));
+							URI newURI = newRequest.getRequestHeader().getURI();
+							if(!oldURI.getHost().equals(newURI.getHost())) {
```

Comment:
This check should also be done for the other URI, the one in the catch block.
Consider extracting a method that creates the URI (returning null if it's not able to), should be easier to read the code and do the host check.